File: pct/photo-coordinates/get-popup-data.py
import os
import json
from exif import Image

directory = "./photos"
outputList = []

def openFile(path):
    with open(path, "rb") as file:
        pic = Image(file)
        readExif(pic)
        

def readExif(pic):
    if pic.has_exif:
        try: 
            print(f"{entry.path} contains EXIF (version {pic.exif_version}) information.")
            print(pic.gps_latitude)
            print(f"Latitude: {pic.gps_latitude} {pic.gps_latitude_ref}")
            print(f"Longitude: {pic.gps_longitude} {pic.gps_longitude_ref}")
            data = parseGPS(pic.gps_latitude, pic.gps_latitude_ref, pic.gps_longitude, pic.gps_longitude_ref, pic.gps_altitude)
            dataObject = createPhotoDataObject(data, entry.path)
            outputList.append(dataObject)
        except AttributeError:
            print(f"Attribute does not exist for {entry.path}")
    else:
        print(f"No EXIF information for {entry.path}")


def parseGPS(lat, lat_ref, lon, lon_ref, alt):
    coordinates = [lat[0] + lat[1]/60 + lat[2]/3600, lon[0] + lon[1]/60 + lon[2]/3600, alt*3.28084] # Convert altitude to feet
    if lat_ref == "S":
        coordinates[0] = coordinates[0]*-1
    if lon_ref == "W":
        coordinates[1] = coordinates[1]*-1
    return(coordinates)


def createPhotoDataObject(data, path):
    # Create dictionary
    obj = {
        "type": "Feature",
        "properties": {
            "description": f"<img src='{path}' width='100%'><p>Altitude = {round(data[2])}ft</p>",
            "icon": "attraction-15"
        },
        "geometry": {
            "type": "Point",
            "coordinates": [data[1], data[0]]
        }
    }
    return(obj)


for entry in os.scandir(directory):
    if (entry.path.endswith(".jpg")
            or entry.path.endswith(".png")) or entry.path.endswith(".JPG") and entry.is_file():
        openFile(entry.path)
    print("\n")

with open("popup-data.js", "w") as f:
    f.write("export const features = ")
    json.dump(outputList, f)
    f.close

File: pct/photo-coordinates/get-popup-data.js
const { readdir } = require("fs/promises");
const fs = require("fs");
var exif = require("exif").ExifImage;

var photos = {}
photos.data = []
var arr = [];
i = 0;

async function run(path) {
    try {
        const files = await readdir(path);


        const data = await photoDataObject(files, path)
            // const obj = await popupObject(files, path);
            // setTimeout(() => { writeOut(obj) }, 5000);
    } catch (err) {
        console.error(err);
    }
}

// Create an object for photo data
const photoDataObject = (files, path) => {
    return new Promise((resolve) => {
        for (const file of files) {
            try {
                new exif({ image: `${path}/${file}` }, function(error, exifData) {
                    if (error) {
                        console.log('Error: ' + error.message);
                    } else {
                        try {
                            latitude = exifData.gps.GPSLatitude;
                            longitude = exifData.gps.GPSLongitude;
                            latLongDecimal = parseGPS(latitude, longitude);
                            exifAltitude = exifData.gps.GPSAltitude; // Meters
                            var obj = {
                                name: file,
                                latitude: latLongDecimal[0],
                                longitude: latLongDecimal[1],
                                altitude: Math.round(exifAltitude * 3.28084) // Feet
                            };
                            console.log(obj);
                            arr.push(obj);
                        } catch (err) {
                            console.log(`Error: ${err} for file ${file}`);
                        }
                    }
                });
            } catch (error) {
                console.log('Error: ' + error.message);
            }
        }
        resolve(arr)
    });
}


// Create object for final popup data
const popupObject = (files) => {
    return new Promise((resolve) => {
        for (const file of files) {
            console.log(file);
            var obj = {
                type: "Feature",
                properties: obj = {
                    // "description": `<img src='${photosDirectory}/${file}' width='100%'><p>Altitude = ${Math.round(altitude)}ft</p>`,
                    "description": `<img src=' width='100%'><p>Altitude =ft</p>`,
                    "icon": "attraction-15"
                },
                geometry: obj = {
                    "type": "Point",
                    "coordinates": [-118.29203 + i, 36.57861333333334 + i]
                }
            }
            photos.data.push(obj);
            i++;
        }
        resolve(photos);
    });
}


// Write object to JSON
function writeOut(object) {
    new Promise((resolve, reject) => {
        fs.writeFile("popup-data.json", JSON.stringify(object), function(err) {
            if (err) {
                return reject(err.message);
            }
            resolve(console.log("File write complete"));
        });
    });
}

function parseGPS(lat, lon) {
    // https://gis.stackexchange.com/questions/136925/how-to-parse-exif-gps-information-to-lat-lng-decimal-numbers
    return [lat[0] + lat[1] / 60 + lat[2] / 3600, -1 * (lon[0] + lon[1] / 60 + lon[2] / 3600)];
}

run("./photos-test");

// console.log("did this go?");

// fs.writeFile("photo-data.json", JSON.stringify(photos), function(err) {
//     if (err) throw err;
//     console.log('complete');
// });